WordPress/src/main/java/org/wordpress/android/networking/menus/MenusRestWPCom.java

```diff
@@ -0,0 +1,162 @@
+    private static final String DELETE_MENU_REST_PATH = "/sites/%s/menus/%s/delete";
+
+    //
+    // JSON keys for fetchAllMenus response object
+    //
+    public static final String MENU_KEY = "menu";
+    public static final String ALL_MENUS_MENUS_KEY = "menus";
+
+    private static final String DELETED_KEY = "deleted";
+
+    private final IMenusDelegate mDelegate;
+
+    public MenusRestWPCom(IMenusDelegate delegate) {
+        if (delegate == null) {
+            throw new IllegalArgumentException("IMenusDelegate cannot be null");
+        }
+        mDelegate = delegate;
+    }
+
+    public boolean createMenu(final MenuModel menu) {
+        // a name string is required for the REST call
+        if (menu == null || TextUtils.isEmpty(menu.name)) return false;
+        String path = formatPath(CREATE_MENU_REST_PATH, null);
+        Map<String, String> params = new HashMap<>();
+        params.put(MENU_NAME_KEY, menu.name);
+        WordPress.getRestClientUtilsV1_1().post(path, params, null, new RestRequest.Listener() {
+            @Override public void onResponse(JSONObject response) {
+                menu.menuId = response.optLong(MENU_ID_KEY);
+                mDelegate.onMenuCreated(HttpURLConnection.HTTP_OK, menu);
+                // TODO: call updateMenu if the menu has any non-default fields
+            }
+        }, new RestRequest.ErrorListener() {
+            @Override public void onErrorResponse(VolleyError error) {
+                mDelegate.onMenuCreated(statusCodeFromVolleyError(error), menu);
+            }
+        });
+        return true;
+    }
+
+    public boolean updateMenu(final MenuModel menu) {
+        if (menu == null) return false;
+        String path = formatPath(MENU_REST_PATH, String.valueOf(menu.menuId));
+        Map<String, String> params = new HashMap<>();
+        params.put(MENU_NAME_KEY, menu.name);
+        WordPress.getRestClientUtilsV1_1().post(path, params, null, new RestRequest.Listener() {
+            @Override public void onResponse(JSONObject response) {
+                MenuModel menu = menuFromJson(response.optJSONObject(MENU_KEY));
+                mDelegate.onMenuUpdated(HttpURLConnection.HTTP_OK, menu);
+            }
+        }, new RestRequest.ErrorListener() {
+            @Override public void onErrorResponse(VolleyError error) {
+                mDelegate.onMenuUpdated(statusCodeFromVolleyError(error), menu);
+            }
+        });
+        return true;
+    }
+
+    public void fetchMenu(long menuId) {
+        String path = formatPath(MENU_REST_PATH, String.valueOf(menuId));
+        Map<String, String> params = new HashMap<>();
+        WordPress.getRestClientUtilsV1_1().get(path, params, null, new RestRequest.Listener() {
+            @Override public void onResponse(JSONObject response) {
+                JSONObject menuObject = response.optJSONObject(MENU_KEY);
+                mDelegate.onMenuReceived(HttpURLConnection.HTTP_OK, menuFromJson(menuObject));
```

Comment:
I have found that when fetching a particular menu id, (at least) 3 things can happen:
- you find the menu
- you don’t find the menu
- there’s a particular case for id=0 as handled by the API

For the first case, all seems good here.
For the second case, the JSON body as returned by the server goes like this:
`{
  "menu": {
    "items": [
      {
        "": false
      }
    ],
    "locations": []
  }
}
`
Good call on making the method `menuFromJson(JSONObject menu)` return `null` in this case, as per the `isValidMenuJson` check. I wonder why the API returns this structure anyway, and a `200 OK`… I would have expected a `404 NOT FOUND` but, I guess this is what we have.

The thing that I don’t like about this, and is addressed in a comment above, is that ambiguity in having `onMenuReceived` with a param of 200 and a null menu object. It’s consistent with the API but it’s not consistent semantically. Again, if `onMenuReceived` is called, it should be only for the case when we are 100% sure a menu has been retrieved, and all other cases should be considered failures, with the reason for failing if known. After all, we asked our MenusRestWPCom class to `fetchMenu`. If it didn’t succeed in doing so, it inevitably _is_ a failure.

For the third case as shown on another line above, the API returns the following JSON body:
`{
  "error": "menu-id",
  "message": "Menu ID must be greater than 0."
}
`
I think this should have a java counterpart to deserialise (maybe an Error class) and handle correctly by the `onErrorResponse` method in the listener interface